File: server-rs/crates/cordy-weixin/src/resolvers.rs
use std::sync::Arc;

use serde::{Deserialize, Serialize};
use sqlx::PgPool;
use uuid::Uuid;

use cordy_channel::InboundMessage;
use cordy_channel_engine::resolvers::{
    AppendParams, AppendResult, BindMediaParams, Deduper, DropReason, EnsureSessionParams,
    IdentityResolver, InstallationResolver, ResolvedIdentity, ResolvedInstallation, ResolverError,
    ResolverSet, SessionBinder,
};
use cordy_channel_engine::session::{
    AppendInput, BindMediaInput, ChatSession, EnsureSessionInput, SessionTitles,
};
use cordy_db::queries::channel::{
    claim_channel_inbound_dedup, get_channel_installation_by_app_id,
    get_channel_user_binding_by_user_id, mark_channel_inbound_dedup_processed,
    merge_channel_chat_session_binding_config, record_channel_inbound_drop,
    release_channel_inbound_dedup,
};
use cordy_db::queries::member::get_member_by_user_and_workspace;

use crate::inbound::WeixinRawEvent;

pub const ORIGIN_WEIXIN_CHAT: &str = "weixin_chat";

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct WeixinBindingConfig {
    pub user_id: String,
    pub context_token_encrypted: String,
}

pub type ContextSealer = dyn Fn(&[u8]) -> anyhow::Result<Vec<u8>> + Send + Sync;

fn raw(message: &InboundMessage) -> anyhow::Result<WeixinRawEvent> {
    serde_json::from_value(message.raw.clone())
        .map_err(|error| anyhow::anyhow!("decode weixin inbound raw: {error}"))
}

struct InstallationResolverImpl {
    pool: PgPool,
}

#[async_trait::async_trait]
impl InstallationResolver for InstallationResolverImpl {
    async fn resolve_installation(
        &self,
        message: &InboundMessage,
    ) -> anyhow::Result<ResolvedInstallation> {
        let raw = raw(message)?;
        let Some(installation) =
            get_channel_installation_by_app_id(&self.pool, crate::TYPE_WEIXIN, &raw.bot_id).await?
        else {
            return Err(ResolverError::InstallationNotFound.into());
        };
        Ok(ResolvedInstallation {
            id: installation.id,
            workspace_id: installation.workspace_id,
            agent_id: installation.agent_id,
            route_revision: 0,
            installer_user_id: installation.installer_user_id,
            active: installation.status == "active",
            platform: Arc::new(installation),
        })
    }
}

struct IdentityResolverImpl {
    pool: PgPool,
}

#[async_trait::async_trait]
impl IdentityResolver for IdentityResolverImpl {
    async fn resolve_sender(
        &self,
        installation: &ResolvedInstallation,
        message: &InboundMessage,
    ) -> anyhow::Result<ResolvedIdentity> {
        let Some(binding) = get_channel_user_binding_by_user_id(
            &self.pool,
            installation.id,
            &message.source.sender_id,
        )
        .await?
        else {
            return Err(ResolverError::SenderUnbound.into());
        };
        if get_member_by_user_and_workspace(
            &self.pool,
            binding.cordy_user_id,
            installation.workspace_id,
        )
        .await?
        .is_none()
        {
            return Err(ResolverError::SenderNotMember.into());
        }
        Ok(ResolvedIdentity {
            user_id: binding.cordy_user_id,
        })
    }
}

struct DeduperImpl {
    pool: PgPool,
}

#[async_trait::async_trait]
impl Deduper for DeduperImpl {
    async fn claim(&self, installation_id: Uuid, message_id: &str) -> anyhow::Result<Uuid> {
        claim_channel_inbound_dedup(&self.pool, installation_id, message_id)
            .await?
            .map(|row| row.claim_token)
            .ok_or_else(|| ResolverError::Duplicate.into())
    }

    async fn mark(&self, installation_id: Uuid, message_id: &str, claim_token: Uuid) {
        let _ = mark_channel_inbound_dedup_processed(
            &self.pool,
            installation_id,
            message_id,
            claim_token,
        )
        .await;
    }

    async fn release(&self, installation_id: Uuid, message_id: &str, claim_token: Uuid) {
        let _ = release_channel_inbound_dedup(&self.pool, installation_id, message_id, claim_token)
            .await;
    }
}

struct SessionBinderImpl {
    pool: PgPool,
    session: ChatSession,
    seal: Arc<ContextSealer>,
}

#[async_trait::async_trait]
impl SessionBinder for SessionBinderImpl {
    async fn ensure_session(&self, params: EnsureSessionParams) -> anyhow::Result<Uuid> {
        let raw = raw(&params.message)?;
        let binding_key = params.message.source.chat_id.clone();
        use base64::Engine as _;
        let context_token_encrypted = base64::engine::general_purpose::STANDARD
            .encode((self.seal)(raw.context_token.as_bytes())?);
        let config = serde_json::to_value(WeixinBindingConfig {
            user_id: params.message.source.sender_id.clone(),
            context_token_encrypted,
        })?;
        let session_id = self
            .session
            .ensure_session(&EnsureSessionInput {
                workspace_id: params.installation.workspace_id,
                agent_id: params.installation.agent_id,
                installation_id: params.installation.id,
                sender: params.sender,
                binding_key: binding_key.clone(),
                binding_config: Some(config.clone()),
                chat_type: params.message.source.chat_type.clone(),
            })
            .await?;
        // `ensure_session` intentionally avoids writes on the hot existing
        // path. iLink's context token changes over time, so merge the latest
        // value after every accepted inbound message for delayed ChatDone.
        merge_channel_chat_session_binding_config(
            &self.pool,
            params.installation.id,
            &binding_key,
            &config,
        )
        .await?;
        Ok(session_id)
    }

    async fn mark_pending_fresh(&self, session_id: Uuid) -> anyhow::Result<()> {
        self.session.mark_pending_fresh(session_id).await
    }

    async fn append_message(&self, params: AppendParams) -> anyhow::Result<AppendResult> {
        self.session
            .append_user_message(&AppendInput {
                session_id: params.session_id,
                sender: params.sender,
                installation_id: params.installation_id,
                body: params.message.text.clone(),
                command_text: params.message.command_text.clone(),
                message_id: params.message.message_id.clone(),
                thread_id: String::new(),
                claim_token: params.claim_token,
                media_pending_seconds: params.media_pending_seconds,
                force_fresh: params.message.force_fresh,
            })
            .await
    }

    async fn bind_media(&self, params: BindMediaParams) -> anyhow::Result<()> {
        self.session
            .bind_media_refs(&BindMediaInput {
                message_id: params.message_id,
                session_id: params.session_id,
                workspace_id: params.workspace_id,
                sender: params.sender,
                issue_id: params.issue_id,
                issue_description_base: params.issue_description_base,
                issue_command_text: params.issue_command_text,
                body: params.body,
                media_refs: params.media_refs,
            })
            .await
    }
}

struct AuditorImpl {
    pool: PgPool,
}

#[async_trait::async_trait]
impl cordy_channel_engine::resolvers::Auditor for AuditorImpl {
    async fn record_drop(
        &self,
        installation_id: Uuid,
        message: &InboundMessage,
        reason: &DropReason,
    ) {
        let _ = record_channel_inbound_drop(
            &self.pool,
            crate::TYPE_WEIXIN,
            "message",
            &reason.0,
            (!installation_id.is_nil()).then_some(installation_id),
            Some(&message.source.chat_id),
            Some(&message.event_id),
            Some(&message.message_id),
            cordy_db::dbid::new_v7(),
        )
        .await;
    }
}

pub fn resolver_set(
    pool: PgPool,
    replier: Option<Arc<dyn cordy_channel_engine::resolvers::OutboundReplier>>,
    seal: Arc<ContextSealer>,
) -> ResolverSet {
    let session = ChatSession::new(
        pool.clone(),
        cordy_channel::Type(crate::TYPE_WEIXIN.to_string()),
        SessionTitles {
            group: "WeChat group".into(),
            direct: "WeChat direct message".into(),
            fallback: "WeChat chat".into(),
        },
    );
    ResolverSet {
        installation: Some(Arc::new(InstallationResolverImpl { pool: pool.clone() })),
        identity: Some(Arc::new(IdentityResolverImpl { pool: pool.clone() })),
        validated: None,
        dedup: Some(Arc::new(DeduperImpl { pool: pool.clone() })),
        session: Some(Arc::new(SessionBinderImpl {
            pool: pool.clone(),
            session,
            seal,
        })),
        media: None,
        audit: Some(Arc::new(AuditorImpl { pool })),
        replier,
        typing: None,
        origin_type: ORIGIN_WEIXIN_CHAT.to_string(),
    }
}

File: server/internal/integrations/weixin/resolvers.go
package weixin

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"strings"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgtype"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/patchbay-ai/patchbay/server/internal/integrations/channel"
	"github.com/patchbay-ai/patchbay/server/internal/integrations/channel/engine"
	db "github.com/patchbay-ai/patchbay/server/pkg/db/generated"
	"github.com/patchbay-ai/patchbay/server/pkg/dbid"
)

type ContextSealer func([]byte) ([]byte, error)

type weixinBindingConfig struct {
	UserID                string `json:"user_id"`
	ContextTokenEncrypted string `json:"context_token_encrypted"`
}

// NewResolverSet wires iLink into the unchanged channel engine. The adapter
// owns only provider-specific routing and context-token storage; session,
// dedup, membership, issue and task behavior remain shared.
func NewResolverSet(q *db.Queries, tx engine.TxStarter, pool *pgxpool.Pool, replier engine.OutboundReplier, seal ContextSealer) engine.ResolverSet {
	set := engine.ResolverSet{
		Installation: &installationResolver{q: q},
		Identity:     &identityResolver{q: q},
		Dedup:        &deduper{q: q},
		Session: &sessionBinder{
			session: engine.NewChatSession(q, tx, TypeWeixin, engine.SessionTitles{
				Group: "Weixin chat", Direct: "Weixin direct message", Fallback: "Weixin chat",
			}),
			pool: pool,
			seal: seal,
		},
		Audit:      &auditor{q: q},
		Replier:    replier,
		OriginType: OriginWeixinChat,
	}
	return set
}

var (
	_ engine.InstallationResolver = (*installationResolver)(nil)
	_ engine.IdentityResolver     = (*identityResolver)(nil)
	_ engine.Deduper              = (*deduper)(nil)
	_ engine.SessionBinder        = (*sessionBinder)(nil)
	_ engine.Auditor              = (*auditor)(nil)
)

func decodeRawEvent(msg channel.InboundMessage) (RawEvent, error) {
	if len(msg.Raw) == 0 {
		return RawEvent{}, errors.New("weixin: inbound Raw is empty")
	}
	var raw RawEvent
	if err := json.Unmarshal(msg.Raw, &raw); err != nil {
		return RawEvent{}, fmt.Errorf("decode weixin inbound Raw: %w", err)
	}
	if strings.TrimSpace(raw.BotID) == "" || strings.TrimSpace(raw.ContextToken) == "" {
		return RawEvent{}, errors.New("weixin: inbound Raw is missing bot or context identity")
	}
	return raw, nil
}

func validDirectSource(msg channel.InboundMessage) bool {
	return msg.Source.ChannelType == TypeWeixin && msg.Source.ChatType == channel.ChatTypeP2P &&
		strings.TrimSpace(msg.Source.ChatID) != "" &&
		strings.TrimSpace(msg.Source.ChatID) == strings.TrimSpace(msg.Source.SenderID)
}

type installationResolver struct{ q *db.Queries }

func (r *installationResolver) ResolveInstallation(ctx context.Context, msg channel.InboundMessage) (engine.ResolvedInstallation, error) {
	if !validDirectSource(msg) {
		return engine.ResolvedInstallation{}, engine.ErrInstallationNotFound
	}
	raw, err := decodeRawEvent(msg)
	if err != nil {
		return engine.ResolvedInstallation{}, err
	}
	if r.q == nil {
		return engine.ResolvedInstallation{}, errors.New("weixin: installation query is not configured")
	}
	inst, err := r.q.GetChannelInstallationByAppID(ctx, db.GetChannelInstallationByAppIDParams{
		ChannelType: string(TypeWeixin), AppID: raw.BotID,
	})
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return engine.ResolvedInstallation{}, engine.ErrInstallationNotFound
		}
		return engine.ResolvedInstallation{}, err
	}
	if inst.ChannelType != string(TypeWeixin) || inst.WorkspaceID.Valid == false || inst.AgentID.Valid == false {
		return engine.ResolvedInstallation{}, engine.ErrInstallationNotFound
	}
	return engine.ResolvedInstallation{
		ID: inst.ID, WorkspaceID: inst.WorkspaceID, AgentID: inst.AgentID,
		InstallerUserID: inst.InstallerUserID, Installed: inst.Status == "installed", Platform: inst,
	}, nil
}

type identityResolver struct{ q *db.Queries }

func (r *identityResolver) ResolveSender(ctx context.Context, inst engine.ResolvedInstallation, msg channel.InboundMessage) (engine.ResolvedIdentity, error) {
	if r.q == nil || !inst.ID.Valid || !inst.WorkspaceID.Valid || !validDirectSource(msg) {
		return engine.ResolvedIdentity{}, engine.ErrSenderUnbound
	}
	binding, err := r.q.GetChannelUserBindingByUserID(ctx, db.GetChannelUserBindingByUserIDParams{
		InstallationID: inst.ID, ChannelUserID: strings.TrimSpace(msg.Source.SenderID),
	})
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return engine.ResolvedIdentity{}, engine.ErrSenderUnbound
		}
		return engine.ResolvedIdentity{}, err
	}
	// The generic binding query is installation-scoped, but these fields are
	// still checked explicitly because the schema has no FK by policy.
	if binding.ChannelType != string(TypeWeixin) || binding.InstallationID != inst.ID || binding.WorkspaceID != inst.WorkspaceID ||
		binding.ChannelUserID != strings.TrimSpace(msg.Source.SenderID) {
		return engine.ResolvedIdentity{}, engine.ErrSenderUnbound
	}
	if _, err := r.q.GetMemberByUserAndWorkspace(ctx, db.GetMemberByUserAndWorkspaceParams{
		UserID: binding.PatchbayUserID, WorkspaceID: inst.WorkspaceID,
	}); err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return engine.ResolvedIdentity{}, engine.ErrSenderNotMember
		}
		return engine.ResolvedIdentity{}, err
	}
	return engine.ResolvedIdentity{UserID: binding.PatchbayUserID}, nil
}

type deduper struct{ q *db.Queries }

func (d *deduper) Claim(ctx context.Context, installationID pgtype.UUID, messageID string) (pgtype.UUID, error) {
	row, err := d.q.ClaimChannelInboundDedup(ctx, db.ClaimChannelInboundDedupParams{
		InstallationID: installationID, MessageID: messageID,
	})
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return pgtype.UUID{}, engine.ErrDuplicate
		}
		return pgtype.UUID{}, err
	}
	return row.ClaimToken, nil
}

func (d *deduper) Mark(ctx context.Context, installationID pgtype.UUID, messageID string, claimToken pgtype.UUID) error {
	_, err := d.q.MarkChannelInboundDedupProcessed(ctx, db.MarkChannelInboundDedupProcessedParams{
		InstallationID: installationID, MessageID: messageID, ClaimToken: claimToken,
	})
	return err
}

func (d *deduper) Release(ctx context.Context, installationID pgtype.UUID, messageID string, claimToken pgtype.UUID) error {
	_, err := d.q.ReleaseChannelInboundDedup(ctx, db.ReleaseChannelInboundDedupParams{
		InstallationID: installationID, MessageID: messageID, ClaimToken: claimToken,
	})
	return err
}

type chatSession interface {
	EnsureSession(context.Context, engine.EnsureSessionInput) (pgtype.UUID, error)
	StartSession(context.Context, engine.StartSessionInput) (engine.StartSessionResult, error)
	MarkPendingFresh(context.Context, pgtype.UUID, string) error
	AppendUserMessage(context.Context, engine.AppendInput) (engine.AppendResult, error)
	BindMediaRefs(context.Context, engine.BindMediaInput) error
}

type sessionBinder struct {
	session chatSession
	pool    *pgxpool.Pool
	seal    ContextSealer
}

func (s *sessionBinder) bindingFor(msg channel.InboundMessage) ([]byte, error) {
	if !validDirectSource(msg) {
		return nil, errors.New("weixin: inbound source is not a fenced direct chat")
	}
	raw, err := decodeRawEvent(msg)
	if err != nil {
		return nil, err
	}
	if s.seal == nil {
		return nil, errors.New("weixin: context token sealer is not configured")
	}
	sealed, err := s.seal([]byte(raw.ContextToken))
	if err != nil {
		return nil, fmt.Errorf("seal weixin context token: %w", err)
	}
	return json.Marshal(weixinBindingConfig{
		UserID:                msg.Source.SenderID,
		ContextTokenEncrypted: base64.StdEncoding.EncodeToString(sealed),
	})
}

func (s *sessionBinder) EnsureSession(ctx context.Context, p engine.EnsureSessionParams) (pgtype.UUID, error) {
	config, err := s.bindingFor(p.Message)
	if err != nil {
		return pgtype.UUID{}, err
	}
	if !validDirectSource(p.Message) {
		return pgtype.UUID{}, errors.New("weixin: only direct chat sessions are supported")
	}
	id, err := s.session.EnsureSession(ctx, engine.EnsureSessionInput{
		WorkspaceID: p.Installation.WorkspaceID, AgentID: p.Installation.AgentID,
		InstallationID: p.Installation.ID, Sender: p.Sender,
		BindingKey: p.Message.Source.ChatID, BindingConfig: config,
		ChatType: p.Message.Source.ChatType,
	})
	if err != nil {
		return pgtype.UUID{}, err
	}
	if s.pool == nil {
		return pgtype.UUID{}, errors.New("weixin: binding config store is not configured")
	}
	if err := mergeBindingConfig(ctx, s.pool, p.Installation.ID, p.Message.Source.ChatID, config); err != nil {
		return pgtype.UUID{}, err
	}
	return id, nil
}

func (s *sessionBinder) StartSession(ctx context.Context, p engine.StartSessionParams) (engine.StartSessionResult, error) {
	config, err := s.bindingFor(p.Message)
	if err != nil {
		return engine.StartSessionResult{}, err
	}
	if !validDirectSource(p.Message) {
		return engine.StartSessionResult{}, errors.New("weixin: only direct chat sessions are supported")
	}
	result, err := s.session.StartSession(ctx, engine.StartSessionInput{
		EnsureSessionInput: engine.EnsureSessionInput{
			WorkspaceID: p.Installation.WorkspaceID, AgentID: p.Installation.AgentID,
			InstallationID: p.Installation.ID, Sender: p.Creator,
			BindingKey: p.Message.Source.ChatID, BindingConfig: config,
			ChatType: p.Message.Source.ChatType,
		},
		Initiator: p.Sender, Body: p.Message.Text, MessageID: p.Message.MessageID,
		ClaimToken: p.ClaimToken, MediaPendingSeconds: p.MediaPendingSeconds,
		PersistMessage: p.PersistMessage, HistoryBoundaryPending: p.HistoryBoundaryPending,
		BeforeCommit: p.BeforeCommit,
	})
	if err != nil {
		return engine.StartSessionResult{}, err
	}
	return result, nil
}

func (s *sessionBinder) MarkPendingFresh(ctx context.Context, sessionID pgtype.UUID, messageID string) error {
	return s.session.MarkPendingFresh(ctx, sessionID, messageID)
}

func (s *sessionBinder) AppendMessage(ctx context.Context, p engine.AppendParams) (engine.AppendResult, error) {
	if !validDirectSource(p.Message) {
		return engine.AppendResult{}, errors.New("weixin: inbound source is not a fenced direct chat")
	}
	commandText := p.Message.CommandText
	if commandText == "" {
		commandText = p.Message.Text
	}
	return s.session.AppendUserMessage(ctx, engine.AppendInput{
		SessionID: p.SessionID, Sender: p.Sender, InstallationID: p.InstallationID,
		Body: p.Message.Text, CommandText: commandText, MessageID: p.Message.MessageID,
		ClaimToken: p.ClaimToken, MediaPendingSeconds: p.MediaPendingSeconds,
		ForceFresh: p.Message.ForceFresh,
	})
}

func (s *sessionBinder) BindMedia(ctx context.Context, p engine.BindMediaParams) (engine.BindMediaResult, error) {
	return engine.BindMediaResult{}, s.session.BindMediaRefs(ctx, engine.BindMediaInput{
		MessageID: p.MessageID, SessionID: p.SessionID, WorkspaceID: p.WorkspaceID,
		Sender: p.Sender, IssueID: p.IssueID, IssueDescriptionBase: p.IssueDescriptionBase,
		IssueCommandText: p.IssueCommandText, Body: p.Body, MediaRefs: p.MediaRefs,
	})
}

type auditor struct{ q *db.Queries }

func (a *auditor) RecordDrop(ctx context.Context, installationID pgtype.UUID, msg channel.InboundMessage, reason engine.DropReason) error {
	raw, _ := decodeRawEvent(msg)
	if a.q == nil {
		return errors.New("weixin: audit query is not configured")
	}
	return a.q.RecordChannelInboundDrop(ctx, db.RecordChannelInboundDropParams{
		ID: dbid.NewV7(), ChannelType: string(TypeWeixin), EventType: raw.EventType,
		DropReason: string(reason), InstallationID: installationID,
		ChannelChatID: nullableText(msg.Source.ChatID), ChannelEventID: nullableText(msg.EventID),
		ChannelMessageID: nullableText(msg.MessageID),
	})
}

func nullableText(value string) pgtype.Text {
	if value == "" {
		return pgtype.Text{}
	}
	return pgtype.Text{String: value, Valid: true}
}
